File: source/nndeploy/inference/inference.cc

#include "nndeploy/inference/inference.h"

#include "nndeploy/base/shape.h"

namespace nndeploy {
namespace inference {

Inference::Inference(base::InferenceType type) {
  type_ = type;
  inference_param_ = createInferenceParam(type);
}

Inference::~Inference() { delete inference_param_; }

base::InferenceType Inference::getInferenceType() { return type_; }

base::Status Inference::setParam(base::Param *param) {
  NNDEPLOY_CHECK_PARAM_NULL_RET_STATUS(param, "param is nullptr");
  return param->copyTo(inference_param_);
}
base::Param *Inference::getParam() {
  return dynamic_cast<base::Param *>(inference_param_);
}

base::ShapeMap Inference::getMinShape() { return inference_param_->min_shape_; }
base::ShapeMap Inference::getOptShape() { return inference_param_->opt_shape_; }
base::ShapeMap Inference::getMaxShape() { return inference_param_->max_shape_; }

int64_t Inference::getMemorySize() {
  NNDEPLOY_LOGI("this api is not implemented");
  return -1;
}
int64_t Inference::getMemorySize(int index) {
  NNDEPLOY_LOGI("this api is not implemented");
  return -1;
}
base::Status Inference::setMemory(device::Buffer *buffer) {
  NNDEPLOY_LOGI("this api is not implemented");
  return base::kStatusCodeOk;
}
base::Status Inference::setMemory(device::Buffer *buffer, int index) {
  NNDEPLOY_LOGI("this api is not implemented");
  return base::kStatusCodeOk;
}

float Inference::getGFLOPs() {
  NNDEPLOY_LOGI("this api is not implemented");
  return 0.0f;
}

bool Inference::isBatch() {
  bool is_batch = false;
  for (auto iter : input_tensors_) {
    device::Tensor *input_tensor = iter.second;
    if (input_tensor->getBatch() > 1) {
      is_batch = true;
      break;
    }
  }
  return is_batch;
}
bool Inference::isShareCommanQueue() { return is_share_command_queue_; }
bool Inference::isInputDynamic() {
  bool is_input_dynamic = false;
  for (auto iter : input_tensors_) {
    device::Tensor *input_tensor = iter.second;
    if (base::isDynamicShape(input_tensor->getShape())) {
      is_input_dynamic = true;
      break;
    }
  }
  if (inference_param_->is_dynamic_shape_) {
    is_input_dynamic = true;
  }
  return is_input_dynamic;
}
bool Inference::isOutputDynamic() {
  bool is_output_dynamic = false;
  for (auto iter : output_tensors_) {
    device::Tensor *output_tensor = iter.second;
    if (base::isDynamicShape(output_tensor->getShape())) {
      is_output_dynamic = true;
      break;
    }
  }
  if (inference_param_->is_dynamic_shape_ || isInputDynamic()) {
    is_output_dynamic = true;
  }
  return is_output_dynamic;
}
bool Inference::canOpInput() {
  bool can_op_input_ = true;
  if (is_share_command_queue_) {
    for (auto iter : input_tensors_) {
      device::Tensor *input_tensor = iter.second;
      if (input_tensor->empty()) {
        can_op_input_ = false;
      }
    }
  } else {
    can_op_input_ = false;
  }
  can_op_input_ = false;
  return can_op_input_;
}
bool Inference::canOpOutput() {
  bool can_op_output = true;
  if (is_share_command_queue_) {
    for (auto iter : output_tensors_) {
      device::Tensor *output_tensor = iter.second;
      if (output_tensor->empty()) {
        can_op_output = false;
      }
    }
  } else {
    can_op_output = false;
  }
  can_op_output = false;
  return can_op_output;
}

int Inference::getNumOfInputTensor() { return input_tensors_.size(); }
int Inference::getNumOfOutputTensor() { return output_tensors_.size(); }

std::string Inference::getInputName(int i) {
  std::vector<std::string> names = getAllInputTensorName();
  if (i < names.size()) {
    return names[i];
  } else {
    return "";
  }
}
std::string Inference::getOutputName(int i) {
  std::vector<std::string> names = getAllOutputTensorName();
  if (i < names.size()) {
    return names[i];
  } else {
    return "";
  }
}

std::vector<std::string> Inference::getAllInputTensorName() {
  std::vector<std::string> input_tensor_names;
  for (auto &tensor : input_tensors_) {
    input_tensor_names.emplace_back(tensor.first);
  }
  return input_tensor_names;
}
std::vector<std::string> Inference::getAllOutputTensorName() {
  std::vector<std::string> output_tensor_names;
  for (auto &tensor : output_tensors_) {
    output_tensor_names.emplace_back(tensor.first);
  }
  return output_tensor_names;
}

base::IntVector Inference::getInputShape(const std::string &name) {
  if (input_tensors_.count(name) > 0) {
    return input_tensors_[name]->getDesc().shape_;
  } else {
    return base::IntVector();
  }
}
base::ShapeMap Inference::getAllInputShape() {
  base::ShapeMap input_shap;
  for (auto &tensor : input_tensors_) {
    input_shap.insert({tensor.first, tensor.second->getDesc().shape_});
  }
  return input_shap;
}

device::TensorDesc Inference::getInputTensorDesc(const std::string &name) {
  if (input_tensors_.count(name) > 0) {
    return input_tensors_[name]->getDesc();
  } else {
    return device::TensorDesc();
  }
}
device::TensorDesc Inference::getOutputTensorDesc(const std::string &name) {
  if (output_tensors_.count(name) > 0) {
    return output_tensors_[name]->getDesc();
  } else {
    return device::TensorDesc();
  }
}

device::TensorDesc Inference::getInputTensorAlignDesc(const std::string &name) {
  if (input_tensors_.count(name) > 0) {
    device::TensorDesc desc = input_tensors_[name]->getDesc();
    if (desc.shape_.size() == 5) {
      if (desc.data_format_ != base::kDataFormatNCDHW &&
          desc.data_format_ != base::kDataFormatNDHWC) {
        desc.data_format_ = base::kDataFormatNCDHW;
      }
    } else if (desc.shape_.size() == 4) {
      if (desc.data_format_ != base::kDataFormatNHWC &&
          desc.data_format_ != base::kDataFormatNCHW) {
        desc.data_format_ = base::kDataFormatNCHW;
      }
    } else if (desc.shape_.size() == 3) {
      if (desc.data_format_ != base::kDataFormatNHW &&
          desc.data_format_ != base::kDataFormatNWC &&
          desc.data_format_ != base::kDataFormatNCW) {
        desc.data_format_ = base::kDataFormatNHW;
      }
    } else if (desc.shape_.size() == 2) {
      if (desc.data_format_ != base::kDataFormatNC) {
        desc.data_format_ = base::kDataFormatNC;
      }
    } else if (desc.shape_.size() == 1) {
      if (desc.data_format_ != base::kDataFormatN) {
        desc.data_format_ = base::kDataFormatN;
      }
    } else {
      desc.data_format_ = base::kDataFormatNotSupport;
    }
    return desc;
  } else {
    return device::TensorDesc();
  }
}
device::TensorDesc Inference::getOutputTensorAlignDesc(
    const std::string &name) {
  if (output_tensors_.count(name) > 0) {
    device::TensorDesc desc = output_tensors_[name]->getDesc();
    if (desc.shape_.size() == 5) {
      if (desc.data_format_ != base::kDataFormatNCDHW &&
          desc.data_format_ != base::kDataFormatNDHWC) {
        desc.data_format_ = base::kDataFormatNCDHW;
      }
    } else if (desc.shape_.size() == 4) {
      if (desc.data_format_ != base::kDataFormatNHWC &&
          desc.data_format_ != base::kDataFormatNCHW) {
        desc.data_format_ = base::kDataFormatNCHW;
      }
    } else if (desc.shape_.size() == 3) {
      if (desc.data_format_ != base::kDataFormatNHW &&
          desc.data_format_ != base::kDataFormatNWC &&
          desc.data_format_ != base::kDataFormatNCW) {
        desc.data_format_ = base::kDataFormatNHW;
      }
    } else if (desc.shape_.size() == 2) {
      if (desc.data_format_ != base::kDataFormatNC) {
        desc.data_format_ = base::kDataFormatNC;
      }
    } else if (desc.shape_.size() == 1) {
      if (desc.data_format_ != base::kDataFormatN) {
        desc.data_format_ = base::kDataFormatN;
      }
    } else {
      desc.data_format_ = base::kDataFormatNotSupport;
    }
    return desc;
  } else {
    return device::TensorDesc();
  }
}

std::map<std::string, device::Tensor *> Inference::getAllInputTensorMap() {
  return input_tensors_;
}
std::map<std::string, device::Tensor *> Inference::getAllOutputTensorMap() {
  return output_tensors_;
}

std::vector<device::Tensor *> Inference::getAllInputTensorVector() {
  std::vector<device::Tensor *> input_tensor;
  for (auto &tensor : input_tensors_) {
    input_tensor.emplace_back(tensor.second);
  }
  return input_tensor;
}
std::vector<device::Tensor *> Inference::getAllOutputTensorVector() {
  std::vector<device::Tensor *> output_tensor;
  for (auto &tensor : output_tensors_) {
    output_tensor.emplace_back(tensor.second);
  }
  return output_tensor;
}

device::Tensor *Inference::getInputTensor(const std::string &name) {
  if (input_tensors_.count(name) > 0) {
    return input_tensors_[name];
  } else {
    return nullptr;
  }
}
device::Tensor *Inference::getOutputTensor(const std::string &name) {
  if (output_tensors_.count(name) > 0) {
    return output_tensors_[name];
  } else {
    return nullptr;
  }
}

base::Status Inference::setInputTensor(const std::string &name,
                                       device::Tensor *input_tensor) {
  base::Status status = base::kStatusCodeOk;

  std::string new_name = "";
  if (!name.empty()) {
    new_name = name;
  } else if (!input_tensor->getName().empty()) {
    new_name = input_tensor->getName();
  } else {
    new_name = getInputName(0);
  }

  if (input_tensors_.count(new_name) > 0) {
    if (input_tensor != input_tensors_[new_name]) {
      external_input_tensors_[new_name] = input_tensor;
    }
  } else {
    NNDEPLOY_LOGI("input_tensor nama: %s not exist!\n", new_name.c_str());
  }

  return status;
}
//
// base::Status Inference::setOutputTensor(const std::string &name,
//                                         device::Tensor *output_tensor) {
//   base::Status status = base::kStatusCodeOk;

//   std::string new_name = "";
//   if (!name.empty()) {
//     new_name = name;
//   } else if (!output_tensor->getName().empty()) {
//     new_name = output_tensor->getName();
//   } else {
//     new_name = getOutputName(0);
//   }

//   if (output_tensors_.count(new_name) > 0) {
//     if (output_tensor != output_tensors_[new_name]) {
//       external_output_tensors_[new_name] = output_tensor;
//     }
//   } else {
//     NNDEPLOY_LOGI("input_tensor nama: %s not exist!\n", new_name.c_str());
//   }

//   return status;
// }

std::map<base::InferenceType, std::shared_ptr<InferenceCreator>>
    &getGlobalInferenceCreatorMap() {
  static std::once_flag once;
  static std::shared_ptr<
      std::map<base::InferenceType, std::shared_ptr<InferenceCreator>>>
      creators;
  std::call_once(once, []() {
    creators.reset(
        new std::map<base::InferenceType, std::shared_ptr<InferenceCreator>>);
  });
  return *creators;
}

Inference *createInference(base::InferenceType type) {
  Inference *temp = nullptr;
  auto &creater_map = getGlobalInferenceCreatorMap();
  if (creater_map.count(type) > 0) {
    temp = creater_map[type]->createInference(type);
  }
  return temp;
}

}  // namespace inference
}  // namespace nndeploy

File: python/nndeploy/inference/inference.py

import nndeploy._nndeploy_internal as _C


import nndeploy.base
import nndeploy.device
import nndeploy.ir
import nndeploy.op

from .inference_param import InferenceParam, InferenceParamCreator, register_inference_param_creator, create_inference_param


# python3 nndeploy/inference/inference.py


class Inference(_C.inference.Inference):
    def __init__(self, type):
        super().__init__(type)
        
    def get_inference_type(self):
        return super().get_inference_type()
    
    def set_param(self, param):
        return super().set_param(param)
    
    def get_param(self):
        return super().get_param()
    
    def get_device_type(self):
        return super().get_device_type()
    
    def set_stream(self, stream):
        return super().set_stream(stream)
    
    def get_stream(self):
        return super().get_stream()
    
    def init(self):
        return super().init()
    
    def deinit(self):
        return super().deinit()
    
    def get_min_shape(self):
        return super().get_min_shape()
    
    def get_opt_shape(self):
        return super().get_opt_shape()
    
    def get_max_shape(self):
        return super().get_max_shape()
    
    def reshape(self, shape_map):
        return super().reshape(shape_map)
    
    def get_memory_size(self):
        return super().get_memory_size()
    
    def set_memory(self, buffer):
        return super().set_memory(buffer)
    
    def get_gflops(self):
        return super().get_gflops()
    
    def is_batch(self):
        return super().is_batch()
    
    def is_share_context(self):
        return super().is_share_context()
    
    def is_share_stream(self):
        return super().is_share_stream()
    
    def is_input_dynamic(self):
        return super().is_input_dynamic()
    
    def is_output_dynamic(self):
        return super().is_output_dynamic()
    
    def can_op_input(self):
        return super().can_op_input()
    
    def can_op_output(self):
        return super().can_op_output()
    
    def get_num_of_input_tensor(self):
        return super().get_num_of_input_tensor()
    
    def get_num_of_output_tensor(self):
        return super().get_num_of_output_tensor()
    
    def get_input_name(self, i=0):
        return super().get_input_name(i)
    
    def get_output_name(self, i=0):
        return super().get_output_name(i)
    
    def get_all_input_tensor_name(self):
        return super().get_all_input_tensor_name()
    
    def get_all_output_tensor_name(self):
        return super().get_all_output_tensor_name()
    
    def get_input_shape(self, name):
        return super().get_input_shape(name)
    
    def get_all_input_shape(self):
        return super().get_all_input_shape()
    
    def get_input_tensor_desc(self, name):
        return super().get_input_tensor_desc(name)
    
    def get_output_tensor_desc(self, name):
        return super().get_output_tensor_desc(name)
    
    def get_input_tensor_align_desc(self, name):
        return super().get_input_tensor_align_desc(name)
    
    def get_output_tensor_align_desc(self, name):
        return super().get_output_tensor_align_desc(name)
    
    def get_all_input_tensor_map(self):
        return super().get_all_input_tensor_map()
    
    def get_all_output_tensor_map(self):
        return super().get_all_output_tensor_map()
    
    def get_all_input_tensor_vector(self):
        return super().get_all_input_tensor_vector()
    
    def get_all_output_tensor_vector(self):
        return super().get_all_output_tensor_vector()
    
    def get_input_tensor(self, name):
        return super().get_input_tensor(name)
    
    def get_output_tensor(self, name):
        return super().get_output_tensor(name)
    
    def set_input_tensor(self, name, input_tensor):
        return super().set_input_tensor(name, input_tensor)
    
    def run(self):
        return super().run()
    
    def get_output_tensor_after_run(self, name, device_type, is_copy, data_format=nndeploy.base.DataFormat.Auto):
        return super().get_output_tensor_after_run(name, device_type, is_copy, data_format)


class InferenceCreator(_C.inference.InferenceCreator):
    def __init__(self):
        super().__init__()

    def create_inference(self, type):
        return super().create_inference(type)



def register_inference_creator(type, creator):
    return _C.inference.register_inference_creator(type, creator)



def create_inference(type):
    return _C.inference.create_inference(type)



